src/protocols/der/encode.c

```diff
@@ -0,0 +1,2014 @@
+{
+	fr_dbuff_t	 our_dbuff = FR_DBUFF(dbuff);
+	fr_pair_t const *vp;
+	char const	*value = NULL;
+	ssize_t		 slen;
+
+	vp = fr_dcursor_current(cursor);
+	if (unlikely(vp == NULL)) {
+		fr_strerror_const("No pair to encode string");
+		return -1;
+	}
+
+	PAIR_VERIFY(vp);
+
+	/*
+	 *	ISO/IEC 8825-1:2021
+	 *	8.23 Encoding for values of the restricted character string types
+	 *		8.23.1 The data value consists of a string of characters from the character set specified in
+	 *			the ASN.1 type definition.
+	 *		8.23.2 Each data value shall be encoded independently of other data values of the same type.
+	 *
+	 *	10.2 String encoding forms
+	 *		For bitstring, octetstring and restricted character string types, the constructed form of
+	 *		encoding shall not be used. (Contrast with 8.23.6.)
+	 *
+	 * 	NOTE:
+	 * 		We DO NOT check for restricted character sets here. This should be done as a separate validation
+	 * 		step. Here we simply trust that administrators have done their job and are providing us with
+	 * 		valid data.
+	 */
+	value = vp->vp_strvalue;
+	slen  = vp->vp_length;
+
+	if (fr_dbuff_in_memcpy(&our_dbuff, value, (size_t)slen) <= 0) {
+		fr_strerror_const("Failed to copy string value to buffer for string type");
+		return -1;
+	}
+
+	return fr_dbuff_set(dbuff, &our_dbuff);
+}
+
+/** Encode the length field of a DER structure
+ *
+ * @param dbuff		The buffer to write the length field to
+ * @param length_start	The start of the length field
+ * @param slen		The length of the structure
+ *
+ * @return		The number of bytes written to the buffer
+ */
+static ssize_t fr_der_encode_len(fr_dbuff_t *dbuff, fr_dbuff_marker_t *length_start, ssize_t slen)
+{
+	fr_dbuff_marker_t value_start;
+	fr_dbuff_t	  value_field;
+	uint8_t		  len_len = 0;
+	ssize_t		  i = 0, our_slen = slen;
+
+	/*
+	 * If the length can fit in a single byte, we don't need to extend the size of the length field
+	 */
+	if (slen <= 0x7f) {
+		fr_dbuff_in(length_start, (uint8_t)slen);
+		return 1;
+	}
+
+	/*
+	 * Calculate the number of bytes needed to encode the length
+	 */
+	while (our_slen > 0) {
+		our_slen >>= 8;
+		len_len++;
+	}
+
+	if (len_len > 0x7f) {
+		fr_strerror_printf("Length %" PRIiPTR " is too large to encode", slen);
+		return -1;
+	}
+
+	value_field = FR_DBUFF(length_start);
+
+	fr_dbuff_set(&value_field, fr_dbuff_current(length_start));
+
+	fr_dbuff_marker(&value_start, &value_field);
+
+	/*
+	 *	Set the dbuff write locaiton to where the new value field will start
+	 */
+	fr_dbuff_set(dbuff, fr_dbuff_current(length_start) + len_len);
+
+	fr_dbuff_move(dbuff, fr_dbuff_ptr(&value_start), slen + 1);
+
+	fr_dbuff_set(dbuff, length_start);
+
+	fr_dbuff_in(dbuff, (uint8_t)(0x80 | len_len));
+
+	for (; i < len_len; i++) {
+		fr_dbuff_in(dbuff, (uint8_t)((slen) >> ((len_len - i - 1) * 8)));
+	}
+
+	fr_dbuff_set(dbuff, fr_dbuff_current(length_start) + len_len + 1 + slen);
```

Comment:
This doesn't seem right?  The function encodes the length, and then skips over the length + the data which will be encoded?

and almost all calls to this function do

```
 ...fr_der_encode_len(&our_dbuff, &length_start, fr_dbuff_behind(&length_start) - 1)
```

except for one...